### PR TITLE
Refactoring - Add formatters

### PR DESCRIPTION
**Pull request by waldyr**
I proposed an enhancement taking almost all methods from the God class Formatter and splitting them among other Formatters classes. Well, here it is and the code speak for itself.

I've extracted methods into classes and specifically colorize went to a module due needs from `Formatter` and `Formatters::BaseFormatter`. Any improvement or comment will be properly addressed specially if it helps me solve my todo quickly. What do you think @gerrywastaken ?

## TODO

- [x] Split methods from Formatter into specific classes.
- [x] Keep common logic together.
- [x] Colorize's destiny
- [x] `options` has an almost global presence and MUST be a better way to deal with that.
- [x] Indentation is not working properly which can be easily recognized by the failing build.

### CODE REVIEW

**Comment by gerrywastaken** [comment]
Wow! That was very quick and impressive. 👍  I've only briefly looked at this as I've spent the entire day moving house and unfortunately had very little sleep. I'll have a better look tomorrow with fresh eyes.

File: lib/awesome_print/formatters/hash_formatter.rb [hunk 1]
```diff
@@ -12,12 +14,13 @@ def initialize(hash, inspector)
       end
 
       def format
-        return "{}" if h == {}
+        return "{}" if hash == {}
```
Comment (gerrywastaken) [inline lib/awesome_print/formatters/hash_formatter.rb]: Thanks for changing some of the single letter variable names

File: lib/awesome_print/formatters/hash_formatter.rb [hunk 2]
```diff
@@ -12,12 +14,13 @@ def initialize(hash, inspector)
       end
```
Comment (gerrywastaken) [inline lib/awesome_print/formatters/hash_formatter.rb]: Thanks for changing some of the single letter variable names

**Comment by gerrywastaken** [comment]
Ok so I'm thinking, with a lot of these public methods which were moved to other classes, maybe it would be a good idea to keep the old method as a simple alias to the new method and add a deprecation message to them.

According to rubygems there were 11.3 million downloads of awesome_print and 3.1 just for the last release. I've also noticed a lot of issues raised in other gems caused by bugs (recently fixed but not yet pushed to rubygems) which were caused by their inclusion of awesome_print as a dependency. I believe it's likely that there are many instances where people have used some of our methods directly.

I think that if we were to got the alias with deprecation route we might save a lot of people a lot of pain. What are your thoughts?

**Comment by waldyr** [comment]
Yeah nothing has changed in terms of interfaces. I've just changed how Formatter do things. Now Formatter delegates to one of the Specific Formatters.

**Comment by waldyr** [comment]
@gerrywastaken Finally I was able to fix the indentation problem... Pheeew 😅 

I also want to know some things before proceeding:
- What do you think about the change in the spec? Do you prefer the old behavior when more entries were printed before collapsing?
- What do you think about Colorize being a module? Do you prefer an object being used by the formatters or it's okay as a module?

**Comment by gerrywastaken** [comment]
@waldyr I'm super sorry for taking so long to get back to you. Nobody told me I was going to be spending the last week building a truck load of IKEA furniture, which ate into all my free time. :-/

> I think refactoring is the best part of Programming

Well I'm glad one of us enjoys it. I do like the end result though. :)

> and you really seem to care about the library. That was enough reasons to do it.

Thanks that makes me happy to hear. :) It makes me sad when I see so many fantastic gems that have been essentially abandoned. It's nobody's fault, but it is something I want to do my part to address.

> but that's a derisory thing.

I just learnt a new word. Thanks!

> What do you think about Colorize being a module? Do you prefer an object being used by the formatters or it's okay as a module?

I'm just glad you separated it out. I'm showing my ignorance here, but why would somebody have a preference about this?

> What do you think about the change in the spec? Do you prefer the old behavior when more entries were printed before collapsing?

OK two things, firstly I believe the failing tests were pointing to a bug.

```
# before this PR
def get_limit_size
  @options[:limit] == true ? DEFAULT_LIMIT_SIZE : @options[:limit]
end

# in this PR
def get_limit_size
  @options[:limit] ? DEFAULT_LIMIT_SIZE : @options[:limit]
end
```
As far as I can make out, the original code wasn't actually checking for a truthy value, but instead checking for it being a boolean **true**. If the limit is set to true, then the default limit is used, otherwise  the limit itself is used. However in the change if you supply a limit the code ends up using the default limit instead, which I don't believe is the desired result.

Changing back to the old version of that method results in the tests passing without that last commit.

I raised one eyebrow when I saw the original code for get_limit_size and I can see why you made that change. While that old version seems to be correct, I think it could really use a comment to explain the logic.

**Comment by waldyr** [comment]
> I'm just glad you separated it out. I'm showing my ignorance here, but why would somebody have a preference about this?

The problem I can see in colorize is that `options` and `inspector` isn't clearly being required. So someone who use the colorize in someplace may not define those methods inspector and options and have problems later.

> I raised one eyebrow when I saw the original code for get_limit_size and I can see why you made that change. While that old version seems to be correct, I think it could really use a comment to explain the logic.

I tried to improve the affordance of the method so anyone who look at it could understand it and don't make the same mistake.

**Comment by gerrywastaken** [comment]
> No worries. I hope you have settled everything up by now. Moving is too stressful.

Cheers. Yup I prefer to avoid it.

> The problem I can see in colorize is that options and inspector isn't clearly being required. So someone who use the colorize in someplace may not define those methods inspector and options and have problems later.

Oh I see what you mean. Well it's still better than what we had. Finding a way of pulling out those dependencies can be a task for another day.

> I tried to improve the affordance of the method so anyone who look at it could understand it and don't make the same mistake.

I like that. It does make things a lot clearer. Thanks for all your hard work on this PR. Merging now. 👍

**Comment by gerrywastaken** [comment]
For anybody reading along, these changes went live in version 1.7.0. A huge thanks to @waldyr who has since been added as a member.

**Pull request merged**